src/meta/meta_bulk_load_ingestion_context.cpp

```diff
@@ -0,0 +1,195 @@
+                address.to_string(),
+                node_ingesting_count,
+                max_node_ingestion_count);
+        return false;
+    }
+
+    auto max_disk_ingestion_count = get_max_disk_ingestion_count(max_node_ingestion_count);
+    if (disk_ingesting_counts[disk_tag] >= max_disk_ingestion_count) {
+        dwarn_f("node[{}] disk[{}] has {} partition executing ingestion, max_count = {}",
+                address.to_string(),
+                disk_tag,
+                disk_ingesting_counts[disk_tag],
+                max_disk_ingestion_count);
+        return false;
+    }
+    return true;
+}
+
+void ingestion_context::node_context::add(const std::string &disk_tag)
+{
+    disk_ingesting_counts[disk_tag]++;
+    node_ingesting_count++;
+}
+
+void ingestion_context::node_context::decrease(const std::string &disk_tag)
+{
+    node_ingesting_count--;
+    disk_ingesting_counts[disk_tag]--;
+}
+
+bool ingestion_context::try_partition_ingestion(const partition_configuration &config,
+                                                const config_context &cc)
+{
+    FAIL_POINT_INJECT_F("ingestion_try_partition_ingestion", [=](string_view) -> bool {
+        auto info = partition_node_info();
+        info.pid = config.pid;
+        _running_partitions[config.pid] = info;
+        return true;
+    });
+    partition_node_info info(config, cc);
+    for (const auto &kv : info.node_disk) {
+        if (!check_node_ingestion(kv.first, kv.second)) {
+            return false;
+        }
+    }
+    add_partition(info);
+    return true;
+}
+
+bool ingestion_context::check_node_ingestion(const rpc_address &node, const std::string &disk_tag)
+{
+    if (_nodes_context.find(node) == _nodes_context.end()) {
+        _nodes_context[node] = node_context(node, disk_tag);
+    }
+    return _nodes_context[node].check_if_add(disk_tag);
+}
+
+void ingestion_context::add_partition(const partition_node_info &info)
+{
+    for (const auto &kv : info.node_disk) {
+        _nodes_context[kv.first].add(kv.second);
+    }
+    _running_partitions[info.pid] = info;
+}
+
+void ingestion_context::remove_partition(const gpid &pid)
+{
+    FAIL_POINT_INJECT_F("ingestion_context_remove_partition",
+                        [=](string_view) { _running_partitions.erase(pid); });
+
+    if (_running_partitions.find(pid) == _running_partitions.end()) {
+        return;
+    }
+    auto &info = _running_partitions[pid];
+    for (const auto &kv : info.node_disk) {
+        _nodes_context[kv.first].decrease(kv.second);
+    }
+    _running_partitions.erase(pid);
+}
+
+uint32_t ingestion_context::get_app_running_count(const uint32_t app_id) const
+{
+    uint32_t running_count = 0;
+    for (const auto &kv : _running_partitions) {
+        if (kv.first.get_app_id() == app_id) {
+            running_count++;
+        }
+    }
+    return running_count;
+}
+
+void ingestion_context::reset_app(const uint32_t app_id)
```

Comment:
`remove_partition` won't do everything if one `gpid` has been removed from `_running_partitions`
Maybe no need to duplicate `gpid` here